streams/src/api/messages.rs

```diff
@@ -241,12 +265,34 @@ impl<'a, T: Send + Sync> MessagesState<'a, T> {
     }
 }
 
+impl<'a, T: Send> From<&'a mut User<T>> for MessagesState<'a, T> {
+    fn from(user: &'a mut User<T>) -> Self {
+        Self {
+            user,
+            ids_stack: Vec::new(),
+            msg_queue: HashMap::new(),
+            stage: VecDeque::new(),
+            successful_round: false,
+            cache: HashMap::new(),
+        }
+    }
+}
+
 impl<'a, T> Messages<'a, T>
 where
     T: for<'b> Transport<'b, Msg = TransportMessage> + Send + Sync,
 {
     pub(crate) fn new(user: &'a mut User<T>) -> Self {
-        let mut state = MessagesState::new(user);
+        let mut state = MessagesState::from(user);
+        Self(Box::pin(async move {
+            let r = state.next().await;
+            (state, r)
+        }))
+    }
+
+    pub(crate) fn new_with_cache(user: &'a mut User<T>, cache: HashMap<MsgId, Message>, start: Option<(Topic, Identifier, usize)>) -> Self {
+        let start = start.into_iter().collect();
+        let mut state = MessagesState::new(user, start, Default::default(), Default::default(), Default::default(), cache);
```

Comment:
Since new is only ever called here, we probably don't need it to have a bunch of Default inputs, that should probably be handled internal for the new() function. You can have it take user, start and cache without the Default::default() calls.